rpc/client.go

```diff
@@ -91,219 +97,224 @@ type Client struct {
 //
 // The Client.Ready channel is closed after the client has connected
 // and completed one successful heartbeat. The Closed channel is
-// closed if the client fails to connect or if the client's Close()
-// method is invoked.
-func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
+// closed if the client's Close() method is invoked.
+func NewClient(addr net.Addr, context *Context) *Client {
 	clientMu.Lock()
+	defer clientMu.Unlock()
+
+	key := util.MakeUnresolvedAddr(addr.Network(), addr.String())
+
 	if !context.DisableCache {
-		if c, ok := clients[addr.String()]; ok {
-			clientMu.Unlock()
+		if c, ok := clients[key]; ok {
 			return c
 		}
 	}
+
+	tlsConfig, err := context.GetClientTLSConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
 	c := &Client{
-		addr:         addr,
-		Ready:        make(chan struct{}),
 		Closed:       make(chan struct{}),
+		addr:         key,
```

Comment:
ah. this was done because the dialer takes strings (rather than a `net.Addr`), and using an UnresolvedAddr allowed me to use its members directly rather than allocating new strings via method calls on the passed in `net.Addr`.